lesson-04-memory-management/mm/xxx.c

```diff
@@ -5,21 +5,142 @@
 #include <linux/pci.h>
 #include <linux/version.h>
 #include <linux/init.h>
+#include <linux/gfp.h>
+#include <linux/sysfs.h>
 
-#define LEN_MSG 160
-static char buf_msg[ LEN_MSG + 1 ] = "Hello from module!\n";
+static size_t g_buf_size = 0;
+static char* g_buf_msg   = 0;
 
+#define MEM_CONFIG_KMALLOC 0
+#define MEM_CONFIG_KMCACHE 1
+
+#define CACHE_NAME "xxx_cache"
+#define CACHE_SIZE 160
+static struct kmem_cache* g_cache = NULL;
+
+static void cache_constructor( void* p )
+{
+   printk( "%s constructs %p(%u)", THIS_MODULE->name, p, (p ? *(char*)p : 0) );
+}
+
+static int g_mem_config = MEM_CONFIG_KMALLOC;
+module_param_named( mem_config, g_mem_config, int, 0 );
```

Comment:
Good idea to use `module_param` for switching implementation, but mixing their code adds a mess.
I'd prefer separate implementations chosen on invocation level or via function table.

Reply:
done